File: renode/peripherals/node_i2c_bridge.py
# Node I2C Bridge — Renode I2C Python Peripheral
#
# Placed on the master machine's I2C0 bus at address 0x08.
# Simulates the node's I2C slave interface so the master firmware can poll it
# in the Renode simulation without requiring true inter-machine I2C wiring.
#
# Protocol (mirrors node firmware):
#   Master writes 0x01  → humidity register selected
#   Master reads  2 bytes → big-endian uint16 = humidity * 100
#
# The humidity values follow the same oscillation as hdc1080.py so the
# master receives plausible, varying readings.
#
# Renode Python peripheral interface (II2CPeripheral):
#   Write(data)
#   Read(count)
#   FinishTransmission()

import math

_reg  = [0x01]
_tick = [0]

def _humidity_x100():
    """Returns humidity * 100 as uint16 (range ~4000–8000 → 40.00–80.00%)."""
    _tick[0] += 1
    rh = 60.0 + 20.0 * math.sin(_tick[0] * 0.15)
    return int(rh * 100) & 0xFFFF

def Read(count):
    if _reg[0] == 0x01:
        val = _humidity_x100()
        return [(val >> 8) & 0xFF, val & 0xFF][:count]
    return [0x00] * count

def Write(data):
    if data:
        _reg[0] = data[0]

def FinishTransmission():
    pass

File: renode/peripherals/NodeI2CBridge.cs
/**
 * NodeI2CBridge.cs
 *
 * Renode I2C peripheral simulating the node's I2C slave interface so the
 * master firmware can poll it within the Renode simulation without requiring
 * true inter-machine I2C wiring.
 *
 * Protocol:
 *   Master writes 0x01, selecting the humidity register.
 *   Master reads 2 bytes, a big-endian uint16 holding humidity * 100.
 *
 * Humidity values follow the same oscillation as HDC1080Device so the
 * master receives plausible, varying readings.
 */

using System;

namespace Antmicro.Renode.Peripherals.I2C
{
    public class NodeI2CBridge : II2CPeripheral
    {
        public NodeI2CBridge()
        {
            Reset();
        }

        public void Reset()
        {
            selectedRegister = 0x01;
            tick = 0;
        }

        public void Write(byte[] data)
        {
            if(data.Length > 0)
            {
                selectedRegister = data[0];
            }
        }

        public byte[] Read(int count = 1)
        {
            if(selectedRegister == 0x01)
            {
                var val = HumidityX100();
                return Truncate(new byte[] {
                    (byte)((val >> 8) & 0xFF),
                    (byte)(val & 0xFF)
                }, count);
            }

            return new byte[count];
        }

        public void FinishTransmission()
        {
        }

        private ushort HumidityX100()
        {
            tick++;
            var rh = 60.0 + 20.0 * Math.Sin(tick * 0.15);
            return (ushort)(int)(rh * 100);
        }

        private static byte[] Truncate(byte[] data, int count)
        {
            if(data.Length <= count)
            {
                return data;
            }

            var result = new byte[count];
            Array.Copy(data, result, count);
            return result;
        }

        private byte selectedRegister;
        private int tick;
    }
}
